From 0ec49292689954633d8235e266a5ac32e9add5ed Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Sun, 28 Jan 2018 12:26:01 +0000
Subject: [PATCH] Correctly integrate with systemd user session

This switches the autostart file out for a systemd user unit, and both
ensures both the service + client portion will only start when the system
is using virtualization.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 Makefile.am                 | 19 ++++++++++---------
 configure.ac                |  2 ++
 data/spice-vdagent.service  | 12 ++++++++++++
 data/spice-vdagentd.service |  1 +
 4 files changed, 25 insertions(+), 9 deletions(-)
 create mode 100644 data/spice-vdagent.service

diff --git a/Makefile.am b/Makefile.am
index 680ef83..3d8c775 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -44,15 +44,6 @@ noinst_HEADERS = src/session-info.h \
                  src/vdagentd-uinput.h \
                  src/vdagentd-xorg-conf.h
 
-xdgautostartdir = $(sysconfdir)/xdg/autostart
-xdgautostart_DATA = $(top_srcdir)/data/spice-vdagent.desktop
-
-gdmautostartdir = $(datadir)/gdm/greeter/autostart
-gdmautostart_DATA = $(top_srcdir)/data/spice-vdagent.desktop
-
-gdmautostart2dir = $(datadir)/gdm/autostart/LoginWindow
-gdmautostart2_DATA = $(top_srcdir)/data/spice-vdagent.desktop
-
 install-data-local:
 	$(mkdir_p) $(DESTDIR)$(localstatedir)/run/spice-vdagentd
 
@@ -67,11 +58,20 @@ systemdunit_DATA = \
 	$(top_srcdir)/data/spice-vdagentd.service \
 	$(top_srcdir)/data/spice-vdagentd.target
 
+systemduserdir = $(SYSTEMDUSERUNITDIR)
+systemduser_DATA = \
+	$(top_srcdir)/data/spice-vdagent.service
+
 udevrulesdir = /lib/udev/rules.d
 udevrules_DATA = $(top_srcdir)/data/70-spice-vdagentd.rules
 
 tmpfilesdir = $(prefix)/lib/tmpfiles.d
 tmpfiles_DATA = $(top_srcdir)/data/tmpfiles.d/spice-vdagentd.conf
+else
+
+xdgautostartdir = $(sysconfdir)/xdg/autostart
+xdgautostart_DATA = $(top_srcdir)/data/spice-vdagent.desktop
+
 endif
 
 manpagedir = $(mandir)/man1
@@ -88,6 +88,7 @@ EXTRA_DIST =					\
 	data/spice-vdagentd.target		\
 	data/tmpfiles.d/spice-vdagentd.conf	\
 	data/xorg.conf.RHEL-5			\
+	data/spice-vdagent.service      \
 	$(NULL)
 
 DISTCHECK_CONFIGURE_FLAGS =			\
diff --git a/configure.ac b/configure.ac
index 2106085..61bdec2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -65,6 +65,8 @@ AC_MSG_RESULT($with_init_script)
 if test "x$init_systemd" = "xyes"; then
   SYSTEMDSYSTEMUNITDIR=`${PKG_CONFIG} systemd --variable=systemdsystemunitdir`
   AC_SUBST(SYSTEMDSYSTEMUNITDIR)
+  SYSTEMDUSERUNITDIR=`${PKG_CONFIG} systemd --variable=systemduserunitdir`
+  AC_SUBST(SYSTEMDUSERUNITDIR)
 fi
 
 AC_ARG_ENABLE([pciaccess],
diff --git a/data/spice-vdagent.service b/data/spice-vdagent.service
new file mode 100644
index 0000000..81e4d2e
--- /dev/null
+++ b/data/spice-vdagent.service
@@ -0,0 +1,12 @@
+[Unit]
+Description=Agent for Spice guests
+After=dbus.service
+ConditionVirtualization=yes
+
+[Service]
+ExecStart=/usr/bin/spice-vdagent
+Type=oneshot
+RemainAfterExit=yes
+
+[Install]
+WantedBy=default.target
diff --git a/data/spice-vdagentd.service b/data/spice-vdagentd.service
index 8c5effe..7bb6005 100644
--- a/data/spice-vdagentd.service
+++ b/data/spice-vdagentd.service
@@ -1,6 +1,7 @@
 [Unit]
 Description=Agent daemon for Spice guests
 After=dbus.target
+ConditionVirtualization=yes
 
 # TODO we should use:
 #Requires=spice-vdagentd.socket
-- 
2.16.1

